include/nbl/builtin/glsl/ext/FFT/fft.glsl

```diff
@@ -0,0 +1,247 @@
+	uint logTwo = 32u - leadingZeroes;
+	
+    nbl_glsl_complex current_values[_NBL_GLSL_EXT_FFT_MAX_ITEMS_PER_THREAD];
+    nbl_glsl_complex shuffled_values[_NBL_GLSL_EXT_FFT_MAX_ITEMS_PER_THREAD];
+
+    // Load Initial Values into Local Mem (bit reversed indices)
+    for(uint t = 0u; t < num_virtual_threads; t++)
+    {
+        uint tid = thread_offset + t * _NBL_GLSL_EXT_FFT_WORKGROUP_SIZE_;
+        uvec3 coords = nbl_glsl_ext_FFT_getCoordinates(tid);
+        uvec3 bitReversedCoords = nbl_glsl_ext_FFT_getBitReversedCoordinates(coords, leadingZeroes);
+
+        current_values[t] = nbl_glsl_ext_FFT_getPaddedData(bitReversedCoords, channel);
+    }
+
+    // For loop for each stage of the FFT (each virtual thread computes 1 buttefly wing)
+	for(uint i = 0u; i < logTwo; ++i) 
+    {
+		uint mask = 1u << i;
+
+        // Data Exchange for virtual threads :
+        // X and Y are seperate to use less shared memory for complex numbers
+        // Get Shuffled Values X for virtual threads
+        for(uint t = 0u; t < num_virtual_threads; t++)
+        {
+            uint tid = thread_offset + t * _NBL_GLSL_EXT_FFT_WORKGROUP_SIZE_;
+            _NBL_GLSL_SCRATCH_SHARED_DEFINED_[tid] = floatBitsToUint(current_values[t].x);
+        }
+        barrier();
+        memoryBarrierShared();
+        for(uint t = 0u; t < num_virtual_threads; t++)
+        {
+            uint tid = thread_offset + t * _NBL_GLSL_EXT_FFT_WORKGROUP_SIZE_;
+            shuffled_values[t].x = uintBitsToFloat(_NBL_GLSL_SCRATCH_SHARED_DEFINED_[tid ^ mask]);
+        }
+
+        // Get Shuffled Values Y for virtual threads
+        for(uint t = 0u; t < num_virtual_threads; t++)
+        {
+            uint tid = thread_offset + t * _NBL_GLSL_EXT_FFT_WORKGROUP_SIZE_;
+            _NBL_GLSL_SCRATCH_SHARED_DEFINED_[tid] = floatBitsToUint(current_values[t].y);
+        }
+        barrier();
+        memoryBarrierShared();
+        for(uint t = 0u; t < num_virtual_threads; t++)
+        {
+            uint tid = thread_offset + t * _NBL_GLSL_EXT_FFT_WORKGROUP_SIZE_;
+            shuffled_values[t].y = uintBitsToFloat(_NBL_GLSL_SCRATCH_SHARED_DEFINED_[tid ^ mask]);
+        }
+
+        // Computation of each virtual thread
+        for(uint t = 0u; t < num_virtual_threads; t++)
+        {
+            uint tid = thread_offset + t * _NBL_GLSL_EXT_FFT_WORKGROUP_SIZE_;
+            nbl_glsl_complex shuffled_value = shuffled_values[t];
+
+            nbl_glsl_complex twiddle = (is_inverse) 
+             ? nbl_glsl_ext_FFT_twiddle(tid, i, logTwo)
+             : nbl_glsl_ext_FFT_twiddleInverse(tid, i, logTwo);
+
+            nbl_glsl_complex this_value = current_values[t];
+
+            if(0u < uint(tid & mask)) {
+                current_values[t] = shuffled_value + nbl_glsl_complex_mul(twiddle, this_value); 
+            } else {
+                current_values[t] = this_value + nbl_glsl_complex_mul(twiddle, shuffled_value); 
+            }
+        }
```

Comment:
only have half the "virtual threads" perform work and do both parts of the butterly each, if you look at the FFT diagram/algorithm the butterflies happen in pairs and the twiddle factor is nearly identical for the pair (so you can compute it only once)

Reply:
totally agreed.
It was originally this way because you wanted me to use ShuffleXOR (in the first version) which would've had no use when a thread computes a full butterfly